processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/MethodSelectors.java

```diff
@@ -36,7 +36,9 @@ public MethodSelectors(TypeUtils typeUtils, ElementUtils elementUtils, TypeFacto
             new InheritanceSelector(),
             new CreateOrUpdateSelector(),
             new SourceRhsSelector(),
-            new FactoryParameterSelector() );
+            new FactoryParameterSelector(),
+            new MostSpecificResultTypeSelector()
```

Comment:
I thought such Selector was there already.. I remember @agudian writing something with distances..

Reply:
I think that you are referring to the `InheritanceSelector`. That is is using distances. However, it is only for the source types. This one is for the return type. I was also contemplating whether to do this in the `TargetTypeSelector`. However, that was a bit too early, since you might have a least specific type with a qualifier